Gems/Atom/RHI/DX12/Code/Source/Platform/Windows/RHI/Device_Windows.cpp

```diff
@@ -222,18 +222,312 @@ namespace AZ
+            case D3D12_AUTO_BREADCRUMB_OP_DECODEFRAME1:
+                return "D3D12_AUTO_BREADCRUMB_OP_DECODEFRAME1";
+            case D3D12_AUTO_BREADCRUMB_OP_SETPROTECTEDRESOURCESESSION:
+                return "D3D12_AUTO_BREADCRUMB_OP_SETPROTECTEDRESOURCESESSION";
+            case D3D12_AUTO_BREADCRUMB_OP_DECODEFRAME2:
+                return "D3D12_AUTO_BREADCRUMB_OP_DECODEFRAME2";
+            case D3D12_AUTO_BREADCRUMB_OP_PROCESSFRAMES1:
+                return "D3D12_AUTO_BREADCRUMB_OP_PROCESSFRAMES1";
+            case D3D12_AUTO_BREADCRUMB_OP_BUILDRAYTRACINGACCELERATIONSTRUCTURE:
+                return "D3D12_AUTO_BREADCRUMB_OP_BUILDRAYTRACINGACCELERATIONSTRUCTURE";
+            case D3D12_AUTO_BREADCRUMB_OP_EMITRAYTRACINGACCELERATIONSTRUCTUREPOSTBUILDINFO:
+                return "D3D12_AUTO_BREADCRUMB_OP_EMITRAYTRACINGACCELERATIONSTRUCTUREPOSTBUILDINFO";
+            case D3D12_AUTO_BREADCRUMB_OP_COPYRAYTRACINGACCELERATIONSTRUCTURE:
+                return "D3D12_AUTO_BREADCRUMB_OP_COPYRAYTRACINGACCELERATIONSTRUCTURE";
+            case D3D12_AUTO_BREADCRUMB_OP_DISPATCHRAYS:
+                return "D3D12_AUTO_BREADCRUMB_OP_DISPATCHRAYS";
+            case D3D12_AUTO_BREADCRUMB_OP_INITIALIZEMETACOMMAND:
+                return "D3D12_AUTO_BREADCRUMB_OP_INITIALIZEMETACOMMAND";
+            case D3D12_AUTO_BREADCRUMB_OP_EXECUTEMETACOMMAND:
+                return "D3D12_AUTO_BREADCRUMB_OP_EXECUTEMETACOMMAND";
+            case D3D12_AUTO_BREADCRUMB_OP_ESTIMATEMOTION:
+                return "D3D12_AUTO_BREADCRUMB_OP_ESTIMATEMOTION";
+            case D3D12_AUTO_BREADCRUMB_OP_RESOLVEMOTIONVECTORHEAP:
+                return "D3D12_AUTO_BREADCRUMB_OP_RESOLVEMOTIONVECTORHEAP";
+            case D3D12_AUTO_BREADCRUMB_OP_SETPIPELINESTATE1:
+                return "D3D12_AUTO_BREADCRUMB_OP_SETPIPELINESTATE1";
+            case D3D12_AUTO_BREADCRUMB_OP_INITIALIZEEXTENSIONCOMMAND:
+                return "D3D12_AUTO_BREADCRUMB_OP_INITIALIZEEXTENSIONCOMMAND";
+            case D3D12_AUTO_BREADCRUMB_OP_EXECUTEEXTENSIONCOMMAND:
+                return "D3D12_AUTO_BREADCRUMB_OP_EXECUTEEXTENSIONCOMMAND";
+            case D3D12_AUTO_BREADCRUMB_OP_DISPATCHMESH:
+                return "D3D12_AUTO_BREADCRUMB_OP_DISPATCHMESH";
+            }
+            return "unkown op";
+        }
+
+        bool Device::AssertSuccess(HRESULT hr)
+        {
+            if (hr == DXGI_ERROR_DEVICE_REMOVED)
+            {
+                OnDeviceRemoved();
+            }
+
+            bool success = SUCCEEDED(hr);
+            AZ_Assert(success, "HRESULT not a success %x", hr);
+            return success;
+        }
+
+        void Device::OnDeviceRemoved()
+        {
+            AZStd::lock_guard<AZStd::mutex> lock(m_onDeviceRemovedMutex);
+
+            // It's possible this function is called many times at same time from different threads.
+            // We only want it be executed once.
+            if (m_onDeviceRemoved)
```

Comment:
Good suggestion. I will change it

Reply:
I actually want the lock since I want the other threads are blocking by this function when it's called. Otherwise the other threads may continue with other code which is not intended.